common/src/main/java/org/corfudb/common/protocol/client/ChannelHandler.java

```diff
@@ -0,0 +1,316 @@
+        if (!channelCf.isDone() && !channelCf.isCompletedExceptionally()) {
+            // throw shutdown exception instead ?
+            channelCf.completeExceptionally(new PeerUnavailable(remoteAddress));
+        }
+        // TODO(Maithem): need to get rid of this future, because clients waiting on this future
+        // need to block till at least 1 connection retry, just fail right away ? queue ops ?
+        this.channelCf = new CompletableFuture<>();
+        // Ideally this should be exponential time off on each retry
+        this.eventLoopGroup.schedule(() -> {
+            log.debug("Retrying to connect to {} in {} ms", remoteAddress, config.getConnectRetryInMs());
+            connect(null);
+        }, config.getConnectRetryInMs(), TimeUnit.MILLISECONDS);
+    }
+
+    public synchronized void close() {
+        requestLock.writeLock().lock();
+        try {
+            if (state == ChannelHandlerState.CLOSED) {
+                // nothing to do
+                return;
+            }
+
+            state = ChannelHandlerState.CLOSED;
+            errorOutAllPendingRequests();
+            CompletableFuture cf = channelCf;
+            if (!cf.isDone() && !cf.isCompletedExceptionally()) {
+                // throw shutdown exception instead ?
+                cf.completeExceptionally(new PeerUnavailable(remoteAddress));
+            }
+
+            Channel currentChannel = channel;
+            if (currentChannel != null) {
+                // TODO(Maithem): will this end up calling channelInactive? need to cancel timer task
+                currentChannel.close().addListener(res -> {
+                    if (!res.isSuccess()) {
+                        log.warn("Failed to close channel for {}", res.cause());
+                    }
+                });
+            }
+        } finally {
+            requestLock.writeLock().unlock();
+        }
+    }
+
+
+    private ChannelInitializer getChannelInitializer() {
+        return new ChannelInitializer() {
+            @Override
+            protected void initChannel(@Nonnull Channel ch) throws Exception {
+                ch.pipeline().addLast(new IdleStateHandler(config.getIdleConnectionTimeoutInMs(),
+                        config.getKeepAlivePeriodInMs(), 0));
+
+                ch.pipeline().addLast(new LengthFieldPrepender(4));
+                ch.pipeline().addLast(new LengthFieldBasedFrameDecoder(Integer.MAX_VALUE,
+                        0, 4, 0,
+                        4));
+
+                /**
+                 ch.pipeline().addLast(new NettyCorfuMessageDecoder());
+                 ch.pipeline().addLast(new NettyCorfuMessageEncoder());
+                 **/
+
+                /**
+                 * //TODO(Maithem): need to implement new handshake logic without netty pipelines
+                 ch.pipeline().addLast(new ClientHandshakeHandler(parameters.getClientId(),
+                 node.getNodeId(), parameters.getHandshakeTimeout()));
+                 **/
```

Comment:
@Maithem Why is it better to implement the new handshake logic outside of the Netty pipeline?

Reply:
I think its cleaner and easier to understand. The current code adds a stage in the pipeline that does the handshake and then removes that stage once the handshake complete. I don't see that as necessary.

Anyways, this can be subjective. But I'm fine with either way.